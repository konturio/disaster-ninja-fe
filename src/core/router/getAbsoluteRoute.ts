import { configRepo } from '~core/config';
import type { AppRoute } from './types';

const trimSlash = (str: string) => {
  // Trim leading and trailing slashes
  return str.replace(/^\/|\/$/g, '');
};

const pathJoin = (...path: Array<string>) => {
  return path
    .map(trimSlash)
    .filter((path) => path.length !== 0)
    .join('/');
};

export const getAbsoluteRoute = (
  slugOrRoute: string | AppRoute,
  base = configRepo.get().baseUrl,
) => {
  if (typeof slugOrRoute === 'string') return '/' + pathJoin(base, slugOrRoute);

  return getAbsoluteRoute(
    slugOrRoute.parentRoute
      ? `${slugOrRoute.parentRoute}/${slugOrRoute.slug}`
      : slugOrRoute.slug,
    base,
  );
};
