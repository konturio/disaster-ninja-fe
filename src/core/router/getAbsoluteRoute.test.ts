import { test, expect } from 'vitest';
import { getAbsoluteRoute } from './getAbsoluteRoute';

const basePathCases = ['/', '/active', '/active/'];
const slugCases = ['', '/', '/slug', '/nested/slug'];

basePathCases.forEach((base) => {
  slugCases.forEach((slug) => {
    test(`Correct url for '${base}' + '${slug}'`, () => {
      expect(getAbsoluteRoute(slug, base)).toMatchSnapshot();
    });
  });
});
