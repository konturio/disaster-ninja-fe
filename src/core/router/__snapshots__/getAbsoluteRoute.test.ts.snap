// Vitest Snapshot v1

exports[`Correct url for '/' + '' 1`] = `"/"`;

exports[`Correct url for '/' + '/' 1`] = `"/"`;

exports[`Correct url for '/' + '/nested/slug' 1`] = `"/nested/slug"`;

exports[`Correct url for '/' + '/slug' 1`] = `"/slug"`;

exports[`Correct url for '/active' + '' 1`] = `"/active"`;

exports[`Correct url for '/active' + '/' 1`] = `"/active"`;

exports[`Correct url for '/active' + '/nested/slug' 1`] = `"/active/nested/slug"`;

exports[`Correct url for '/active' + '/slug' 1`] = `"/active/slug"`;

exports[`Correct url for '/active/' + '' 1`] = `"/active"`;

exports[`Correct url for '/active/' + '/' 1`] = `"/active"`;

exports[`Correct url for '/active/' + '/nested/slug' 1`] = `"/active/nested/slug"`;

exports[`Correct url for '/active/' + '/slug' 1`] = `"/active/slug"`;
